yarn-project/simulator/src/avm/journal/public_storage.ts

```diff
@@ -0,0 +1,149 @@
+
+  constructor(hostPublicStorage: PublicStateDB, parent?: PublicStorage) {
+    this.hostPublicStorage = hostPublicStorage;
+    this.parentCache = parent?.cache;
+    this.cache = new PublicStorageCache();
+  }
+
+  /**
+   * Get the pending storage.
+   */
+  public getCache() {
+    return this.cache;
+  }
+
+  /**
+   * Read a value from storage.
+   * 1. Check cache.
+   * 2. Check parent's cache.
+   * 3. Fall back to the host state.
+   * 4. Not found! Value has never been written to before. Flag it as non-existent and return value zero.
+   *
+   * @param storageAddress - the address of the contract whose storage is being read from
+   * @param slot - the slot in the contract's storage being read from
+   * @returns exists: whether the slot has EVER been written to before, value: the latest value written to slot, or 0 if never written to before
+   */
+  public async read(storageAddress: Fr, slot: Fr): Promise<[/*exists=*/ boolean, /*value=*/ Fr]> {
+    // First try check this storage cache
+    let value = this.cache.read(storageAddress, slot);
+    // Then try parent's storage cache (if it exists / written to earlier in this TX)
+    if (!value && this.parentCache) {
+      value = this.parentCache?.read(storageAddress, slot);
+    }
+    // Finally try the host's Aztec state (a trip to the database)
+    if (!value) {
+      value = await this.hostPublicStorage.storageRead(storageAddress, slot);
+    }
+    // if value is undefined, that means this slot has never been written to!
+    const exists = value !== undefined;
+    const valueOrZero = exists ? value : Fr.ZERO;
+    return Promise.resolve([exists, valueOrZero]);
+  }
+
+  /**
+   * Stage a storage write.
+   *
+   * @param storageAddress - the address of the contract whose storage is being written to
+   * @param slot - the slot in the contract's storage being written to
+   * @param value - the value being written to the slot
+   */
+  public write(storageAddress: Fr, key: Fr, value: Fr) {
+    this.cache.write(storageAddress, key, value);
+  }
+
+  /**
+   * Merges another PublicStorage's cache (pending writes) into this one.
+   *
+   * @param incomingPublicStorage - the incoming public storage to merge into this instance's
+   */
+  public acceptAndMerge(incomingPublicStorage: PublicStorage) {
+    this.cache.acceptAndMerge(incomingPublicStorage.cache);
+  }
+}
+
+/**
+ * A class to cache writes to public storage during a contract call's AVM simulation.
+ * "Writes" update a map, "reads" check that map or return undefined.
+ * An instance of this class can merge another instance's staged writes into its own.
+ */
+class PublicStorageCache {
+  /**
+   * Map for staging storage writes.
+   * One inner-map per contract storage address,
+   * mapping storage slot to latest staged write value.
+   */
+  public cachePerContract: Map<bigint, Map<bigint, Fr>> = new Map();
+  // FIXME: storage ^ should be private, but its value is used in tests for "currentStorageValue"
```

Comment:
We shouldn't really need to expose this as public just so tests can check current values 🤔